### PR TITLE
Basic authentication

### PR DESCRIPTION
**Pull request by sandylcruz**
In this section, I implemented basic backend authentication. I created a users and sessions controller, and I updated the routes.

### CODE REVIEW

File: backend/app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+require 'bcrypt'
+require 'securerandom'
+
+# Create user model
+class User < ApplicationRecord
+  attr_reader :password
+
+  validates :username, presence: true, uniqueness: true
+  validates :email, presence: true, uniqueness: true
+  validates :password_digest, :session_token, presence: true
```
Comment (louisscruz) [inline backend/app/models/user.rb:12]: Uh oh. Session token should be unique too. If it weren’t, I could hijack who the current user is if I can manage to get a two users to have the same session token

File: backend/app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  validates :password, length: { minimum: 6 }, allow_nil: true
+
+  after_initialize :ensure_session_token
+
+  has_many :favorites,
+           primary_key: :id,
+           foreign_key: :user_id
+
+  has_many :reservations,
```
Comment (louisscruz) [inline backend/app/models/user.rb:21]: I don’t think any of the has_many associations are relevant to this PR. remove please

File: backend/app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def password=(password)
+    @password = password
+    self.password_digest = BCrypt::Password.create(password)
+  end
+
+  def reset_session_token!
+    self.session_token = SecureRandom.urlsafe_base64
+    save
```
Comment (louisscruz) [inline backend/app/models/user.rb:47]: Should this be save! ?

File: backend/app/views/static_pages/root.html.erb [hunk 4]
```diff
@@ -1 +1,6 @@
-<div id="root"></div>
+<div id="root">
+<h1>React is not working</h1>
```
Comment (louisscruz) [inline backend/app/views/static_pages/root.html.erb:2]: Maybe a friendlier message here... if users are on a slow internet connection or slow cpu, they will actually see this message for an extended period of time while the JavaScript assets are downloaded and then executed. Common convention is to place “loading...” here

File: backend/config/application.rb [hunk 5]
```diff
@@ -13,6 +13,11 @@ module BrokenTableServer
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.1
+    
+    initializer(:remove_action_mailbox_and_activestorage_routes, after: :add_routing_paths) { |app|
```
Comment (louisscruz) [inline backend/config/application.rb:17]: What do all these changes do?

File: backend/config/routes.rb [hunk 6]
```diff
@@ -2,6 +2,9 @@
 
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
-
+  namespace :api, defaults: { format: :json } do
+    resources :users, only: %i[create show update destroy]
```
Comment (louisscruz) [inline backend/config/routes.rb:6]: Remove update for now

**Review by louisscruz — CHANGES_REQUESTED** [review]
For now we should probably just aim for user new, create, show, destroy. We can add the update at a later time when it’s needed. Can you remove any update code/comments?

File: backend/config/application.rb (hunk 5 above)
Comment (sandylcruz) [reply backend/config/application.rb:17]: Rails automatically adds a bunch of unnecessary routes. This change is to remove the default routes that aren't appropriate.

File: backend/app/controllers/application_controller.rb [hunk 7]
```diff
@@ -1,4 +1,30 @@
 # frozen_string_literal: true
 
+# This is the ApplicationController
 class ApplicationController < ActionController::Base
+  helper_method :current_user, :logged_in?
+
+  def login!(user)
+    session[:session_token] = user.session_token
```
[the remark this answers is not in the document]
Comment (sandylcruz) [reply backend/app/controllers/application_controller.rb:8]: Okay, will reset session token upon login as well.

File: backend/app/models/user.rb (hunk 1 above)
Comment (sandylcruz) [reply backend/app/models/user.rb:12]: Okay, will change session token to be unique.

File: backend/app/models/user.rb (hunk 3 above)
Comment (sandylcruz) [reply backend/app/models/user.rb:47]: Yes it should be save! Changed it.

File: backend/app/views/static_pages/root.html.erb (hunk 4 above)
Comment (sandylcruz) [reply backend/app/views/static_pages/root.html.erb:2]: That makes sense. Have changed it.

File: backend/config/routes.rb (hunk 6 above)
Comment (sandylcruz) [reply backend/config/routes.rb:6]: Okay, have removed update from users' routes.

File: backend/app/controllers/api/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      @user = User.find_by(params[:id])
+
+      if @user
+        render :show
+      else
+        render json @user.errors.full_messages, status: 404
+      end
+    end
+
+    # def update
```
[the remark this answers is not in the document]
Comment (sandylcruz) [reply backend/app/controllers/api/users_controller.rb:33]: Okay have deleted comments.

File: backend/app/models/user.rb (hunk 2 above)
Comment (sandylcruz) [reply backend/app/models/user.rb:21]: Okay, have removed associations

File: backend/app/controllers/api/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  # This is the SessionsController
+  class SessionsController < ApplicationController
+    skip_before_action :verify_authenticity_token
+
+    def new
+      render :new
+    end
+
+    def create
+      @user.find_by_credentials(params[:user][:username], params[:user][:password])
```
[the remark this answers is not in the document]
Comment (sandylcruz) [reply backend/app/controllers/api/sessions_controller.rb:13]: Yes, @user should be User. Have changed it.

File: backend/db/migrate/20210406201150_add_users_table.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+      t.string :username, null: false
+      t.string :email, null: false
+      t.string :password_digest, null: false
+      t.string :session_token, null: false
+      t.datetime :created_at, null: false
+      t.datetime :updated_at, null: false
+    end
+
+    add_index :users, :username, unique: true
+    add_index :users, :email, unique: true
```
Comment (louisscruz) [inline backend/db/migrate/20210406201150_add_users_table.rb:16]: We need an index for session token, and it should be unique

File: backend/db/migrate/20210406222303_remove_null_from_session_token.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+# Remove null constraint on session token
+class RemoveNullFromSessionToken < ActiveRecord::Migration[6.1]
+  def change
+    change_column :users, :session_token, :string, null: true
```
Comment (louisscruz) [inline backend/db/migrate/20210406222303_remove_null_from_session_token.rb:6]: Roll this into the previous migration

**Review by louisscruz — CHANGES_REQUESTED** [review]

**Review by louisscruz — APPROVED** [review]

**Pull request merged**